### ISSUE TITLE
Support for adding a specified route

### ISSUE THREAD

**Issue by Ape** [adrienverge/openfortivpn#371]
I would like to add a custom route on the ppp interface after the VPN connection has been established. I can currently manually run `sudo ip route add 1.2.3.4/32 via 10.20.30.40 dev ppp0` to do this. But it would be a lot easier if openfortivpn did that for me when the interface is created and/or there is a working VPN connection. This way it wouldn't require manual intervention everytime my systems boots (and starts openfortivpn) or after VPN reconnect.

I would suggest something like `--add-route 1.2.3.4/32` which would automatically determine the correct gateway address and interface name from the VPN connection.

I would like to avoid `--set-routes` and `--half-internet-routes` because I have other networking and VPN connections on the same system.

**Comment by DimitriPapadopoulos** [comment]
Duplicate of #22 as far as i can see. See [_pppd_ ip-up scripts](https://www.tldp.org/HOWTO/PPP-HOWTO/x1455.html) and option `--pppd-ipparam`.

Please do not hesitate reopen if that doesn't help.

**Comment by thackel** [comment]
I have similar "problems".

As a "normal" user i don't want to modify/create any pppd/ifup/down scripts on a system global scale.
In the future it might be possible to let the Networkmanager do this like i want, but today at least the KDE version does not support any OTP dialog.

For now i created a script which starts and forks openfortivpn and then adds the routes as desired.
So a one-shot like the OP requested would be nice.

**Comment by DimitriPapadopoulos** [comment]
I believe `--add-route 1.2.3.4/32` would only address part of the requests we have seen. We cannot possibly add multiple options for each different need.

Also I'm not certain it's (currently) possible to clean up the modified route from _openfortivpn_, after having forked _pppd_. Perhaps this is best left to _pppd_. Unfortunately [the privileged way to interact with _pppd_ seems to be through scripts in `/etc/ppp`](https://ppp.samba.org/pppd.html#sect13).

For the above reasons I still believe such needs would be better handled by scripts, either scripts run by _pppd_ or (at the other end) scripts that fork _openfortivpn_, as suggested above. Hell, I could even agree to a script directly run by _openfortivpn_ if there is no other easy way - but then again we cannot easily clean up when _pppd_ exits.

You are also raising the question how to help _normal_ users. Some ideas, probably not perfect:
* write documentation,
* provide sample scripts,
* have NetworkManager generate scripts,
* have _pppd_ run a script outside `/etc/ppp` (but how?),
* or as a last resort even have _openfortivpn_ run a startup script.

And finally why do you need to set up a route manually? Doesn't the gateway provide proper routing instructions? Can you provide a more detailed use case?

**Comment by DimitriPapadopoulos** [comment]
Also I'm not certain a user who runs multiple VPN concurrently with multiple network connections qualifies as _normal_ (with all due respect of course :-).

Such users will probably require different and complex networking options that seem best left to scripting.

**Comment by DimitriPapadopoulos** [comment]
Finally please note this patch that will notify the _pppd_ tunnel is up using _systemd_.

**Comment by DimitriPapadopoulos** [comment]
Additional note: Notifying that the _pppd_ tunnel is up may be best left to _pppd_ itself.

**Comment by DimitriPapadopoulos** [comment]
One last suggestion: It could be up to the packager to have `/etc/ppp/ip-up` start a script found in `~/.ppp/` or elsewhere. But then there are probably associated security issues...

I realize I'm desperately trying to have _pppd_ run your startup scripts instead of _openfortivpn_. Am I completely wrong?

There are many issues at hand:
1. Detect/notify the _pppd_ tunnel is up (or down for the purpose of cleaning up). I believe this is best left to _pppd_ itself. Unfortunately it seems there is no easy way to have _pppd_ notify the IPv4 tunnel is up (and perhaps IPv6 tunnel some day). Ideas?
2. Supposing there is a way to get notified of _pppd_ events, which of _pppd_ or _openfortivpn_ should run startup commands when the _pppd_ tunnel is up (respectively cleanup commands when the _pppd_tunnel is down)? Ideas? Can we easily perform cleanup from _openfortivpn_ when _pppd_ is down?
3. Because the spectrum of functions to be run at startup is infinite, I believe in scripting instead of constantly adding new options to _openfortivpn_. The form taken by scripting very much depends on whether _openfortivpn_ or _pppd_ run the startup commands. Opinions?

More opinions and ideas?

**Comment by DimitriPapadopoulos** [comment]
After reading [this thread](https://news.ycombinator.com/item?id=8749514) I am convinced one of the most elegant solutions would be to have _pppd_ emit _systemd_/_D-Bus_ notifications.

Either ask the _pppd_ maintainers or write a _pppd_ [plugin](https://ftp.samba.org/pub/unpacked/ppp/PLUGINS).

**Comment by feugy** [comment]
@thackel, for what it worth, here is how I've configured pppd to implement selecting routing through openfortivpn:
1. run _openfortivpn_ with `no-route` flag, or add `set-routes = 0` in your configuration file
2. create a ppp script when link is up: `sudo touch /etc/ppp/ip-up.d/fortivpn`
3. add your own routes. I've decided to look up domains by their name:
   ```
   #!/bin/bash
   #
   # Whitelist here all domains that need to go through FortiVPN
   # Domains are separated by a space
   #
   domains='example.com example.fr'
   
   let ip
   for domain in $domains; do
     ip=`dig +short $domain | tail -n1`
     route add $ip dev ppp0
   done
   ```
4. check the created routes:
   ```
   $ route
   Kernel IP routing table
   Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
   default         192.168.10.254  0.0.0.0         UG    600    0        0 wlp3s0
   one.one.one.one *               255.255.255.255 UH    0      0        0 ppp0
   IP-OF-FIRST-DOMAIN *               255.255.255.255 UH    0      0        0 ppp0
   IP-OF-SECOND-DOMAIN *               255.255.255.255 UH    0      0        0 ppp0
   link-local      *               255.255.0.0     U     1000   0        0 br-cc6b09fa8986
   172.16.238.0    *               255.255.255.0   U     0      0        0 br-30ec352ff3e1
   172.17.0.0      *               255.255.0.0     U     0      0        0 docker0
   172.18.0.0      *               255.255.0.0     U     0      0        0 br-44b22a74e13d
   172.19.0.0      *               255.255.0.0     U     0      0        0 br-cc6b09fa8986
   172.20.0.0      *               255.255.0.0     U     0      0        0 br-2718c069c2ed
   192.168.10.0    *               255.255.255.0   U     600    0        0 wlp3s0
   ```
4. check your IP: https://duckduckgo.com/?q=what%27s+my+ip&t=lm&atb=v110-5_h&ia=answer
   It should be your regular IP, while you should have access to white-listed domains through the VPN.

Hop it helps!

**Comment by Eerovil** [comment]
@feugy 
Wow, thanks for the detailed guide!

I gotta ask, how would you add a route for a specific ip (i.e. no domain available)?
There is info about that in the route manual but I couldn't quite grasp it just yet.

**Comment by mrbaseman** [comment]
@DimitriPapadopoulos should we put this on the [wiki](https://github.com/adrienverge/openfortivpn/wiki)? (I have just discovered it recently. I guess you have started it in the context of [troubleshooting](https://github.com/adrienverge/openfortivpn/issues/362#issuecomment-415000855).) We have also thought about putting together an FAQ guide on the wiki in the context of starting _openfortivpn_ [as a service](https://github.com/adrienverge/openfortivpn/issues/229#issuecomment-395957015) and about [respawning](https://github.com/adrienverge/openfortivpn/pull/190#issuecomment-335575984) it.

**Comment by DimitriPapadopoulos** [comment]
Sure, let's improve the wiki!

**Comment by mrbaseman** [comment]
I have done so, copied a few parts from the issues to the wiki. The sections can still be improved, but at least the topics that we have identified as important get more attention than in closed issues

**Comment by feugy** [comment]
Hi @Eerovil, sorry for the delay.

Adding route for specific IP is actually simpler than for domains:
```sh
#!/bin/bash
#
# Whitelist here all domains that need to go through FortiVPN
# Domains are separated by a space
#
domains='example.com example.fr'
ips='192.168.0.15'

let resolved
for domain in $domains; do
  resolved=`dig +short $domain | tail -n1`
  ips="$ips $resolved"
done

for ip in $ips; do
  echo "route add $ip dev ppp0"
done
```

**Comment by mrbaseman** [comment]
thanks @feugy I have updated the example on the [wiki](https://github.com/adrienverge/openfortivpn/wiki) as well

**Comment by mrbaseman** [comment]
I think we can close this issue now. An answer has been provided and we have created the wiki so that users who might not check closed issues can easily find the instructions

**Comment by rohanrajpal** [comment]
Thanks @feugy for the solution. I went ahead and write a blog for any beginner to be able to do this.
https://rohanrajpal.github.io/workfromhome/vpn/networking/selective/routing/systemd/2020/04/25/Selective-network-routing.html
Let me know if I have missed giving credit to this solution or the wiki. I'll update it ASAP.

**Comment by DimitriPapadopoulos** [comment]
@rohanrajpal There's already a _systemd_ service file here:
https://github.com/adrienverge/openfortivpn/blob/master/lib/systemd/system/openfortivpn%40.service.in

Since I'm new to these _systemd_ service files, I'm interested in any feedback you may have. I have seen a few differences between our service files:
* `Type`: `simple` or `idle`, is `idle` better?
* In your case aren't these contradictory?
  ```
   After=[...] multi-user.target
   
   WantedBy=multi-user.target
   ```
* I had initially left this one out, it's probably a good idea to have it so I'll add it:
  ```
  Documentation=man:openfortivpn(1)
  ```
* I think `User=root` and `KillSignal=SIGTERM` are the [default](https://www.freedesktop.org/software/systemd/man/systemd.kill.html#KillSignal=) so I have left them out. Happy to change that in our own service file if required.

**Comment by rohanrajpal** [comment]
Hey, even I'm new to systemd services. Thanks for pointing these things out, I learned a few new things

> @rohanrajpal There's already a _systemd_ service file here:
> https://github.com/adrienverge/openfortivpn/blob/master/lib/systemd/system/openfortivpn%40.service.in
> 
> Since I'm new to these _systemd_ service files, I'm interested in any feedback you may have. I have seen a few differences between our service files:
> 
>     * `Type`: `simple` or `idle`, is `idle` better?
I prefer idle because the service waits until all active jobs are dispatched, I was just trying to avoid any conflict. However, as we have set `After`, I don't think  `simple` should cause any problem. I referred to this [link](https://www.freedesktop.org/software/systemd/man/systemd.service.html#Type=).
> 
>     * In your case aren't these contradictory?
>       ```
>        After=[...] multi-user.target
>        
>        WantedBy=multi-user.target
>       ```
Oh, my bad, I'll update this in my service.
> 
>     * I had initially left this one out, it's probably a good idea to have it so I'll add it:
>       ```
>       Documentation=man:openfortivpn(1)
>       ```
great
> 
>     * I think `User=root` and `KillSignal=SIGTERM` are the [default](https://www.freedesktop.org/software/systemd/man/systemd.kill.html#KillSignal=) so I have left them out. Happy to change that in our own service file if required.
Oh yes, I missed this. I will remove them from mine as well.

**Comment by yeya** [comment]
How this script can be used when I use several instances of openfortivpn in parallel?
It will be `ppp0` only in the first instance. The interface name will change each time.

**Comment by stevenwjy** [comment]
Hello there, I encountered a similar issue with #291 and tried to use the solution described in this thread and [article](https://rohanrajpal.github.io/workfromhome/vpn/networking/selective/routing/systemd/2020/04/25/Selective-network-routing.html) (thanks @rohanrajpal!) to add the route manually. I have managed to run the `openfortivpn` successfully and add the ip to my routing table. However, my ssh still hangs as if it is unable to connect to the port 22 of the given ip (which needs the vpn to connect). Any tips on how to check whether my routing works and things that I could potentially miss when trying to connect?

**Comment by mrbaseman** [comment]
@stevenwjy simply adding a route does not necessarily give you ssh access to that IP. Usually, the Fortigate Appliance pushes the routes to those IPs / networks for which a firewall rule is configured that allows your account to access them. 
Adding a route manually can help in some situations, e.g. when you are dealing with several VPNs or the machine that runs _openfortivpn_ has several interfaces or a complex routing table for some other reasons.

**Comment by ayaka14732** [comment]
On Arch Linux, the proposed method does not work.

To make it work, the file name `/etc/ppp/ip-up.d/fortivpn` should be `/etc/ppp/ip-up.d/fortivpn.sh`.

This is because, according to `/etc/ip-{up,down}`, script names in `/etc/ip-{up,down}.d` need to end with `.sh`. (See [thread](https://bbs.archlinux.org/viewtopic.php?pid=614465#p614465))

We can edit the wiki page accordingly.

**Comment by DimitriPapadopoulos** [comment]
@deadalnix Please open a separate issue on an Arch Linux site. This doesn't look like it's directly related to openfortivpn.

**Comment by deadalnix** [comment]
It's not an arch problem. Sure technically it is not an issue with openforti vpn, but is it really? People keep having problems with the advised solution,a nd because the whole thign happens through openfortivpn, debuggability is really poor.

**Comment by DimitriPapadopoulos** [comment]
OK, but please open a new ticket for new issues. Not certain about "people keep having problems with the advised solution". Please provide logs.

**Comment by tomaszkubacki** [comment]
just wanted to mention I created simplified version of adding custom routes via config file simply by passing custom routes array through **pppd-ipparam** details here:
https://github.com/adrienverge/openfortivpn/wiki#method-one---short-version---pass-routes-by-pppd-ipparam

**Comment by Ryushin** [comment]
I wanted to add to this issue that this solution no longer worked after upgrading to Debian Trixie due to #1134.

I added a line to /etc/ppp/ip-up.d/openfortivpn to delete the bad route:

```
#!/bin/sh -e   

# $PPP_IFACE should also be available with $1, if needed.
logger "forticlient connection is up $PPP_IFACE connection ip routes: ${6}"
IPS="${6}"
for IP in ${IPS}
do
	ip route add "${IP}" dev "${PPP_IFACE}"
done
	ip route del "${PPP_REMOTE}" dev "${PPP_IFACE}"
```

The Wiki might need to be changed.
